### PR TITLE
Support CALDERA API client auto-generation

### PR DESCRIPTION
**Pull request by daw1012345**

### CODE REVIEW

**Comment by MrDurion** [comment]
The modules imported in the generated code are not in the "required modules" yet (go.mod). Should this be added or is there a better convention to fix that?

**Comment by MrDurion** [comment]
I think we should not add any 'dependencies' for generating the code (being able to execute `bash` and `python` scripts), or at least limit it to a single, non platform-specific method. 

Would doing a quick docker run with a volume to the correct place be an option?

**Comment by daw1012345** [comment]
> I think we should not add any 'dependencies' for generating the code (being able to execute `bash` and `python` scripts), or at least limit it to a single, non platform-specific method.
> 
> Would doing a quick docker run with a volume to the correct place be an option?

IMO this is going too far. I can't think of a use case where the **developer** wouldn't have access to bash + python3. The python dependency is also a temporary workaround until caldera fixes their API. The script already spawns an instance of CALDERA in Docker, doing it inside another container would just complicate things with no clear benefit. No special python libraries or packages are used either, just vanilla python and bash.

File: .gitignore [hunk 1]
```diff
@@ -22,3 +22,4 @@ docs/package-lock.json
 docs/.hugo_build.lock
 **.hugo_build.lock
 
+caldera_api*.json
```
Comment (mpboom) [inline .gitignore]: missing newline 😢

File: internal/capability/caldera/api/client/abilities/abilities_client.go [hunk 2]
```diff
@@ -0,0 +1,401 @@
+// Code generated by go-swagger; DO NOT EDIT.
```
Comment (mpboom) [inline internal/capability/caldera/api/client/abilities/abilities_client.go:1]: ye, this generated code should be a separate commit from the scripts introduced to build it

File: build_caldera_api.sh [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+PROJ_PATH="$PWD"
+
+
+CALDERA_VERSION="e4a4b9a320e707e58e69e3213ca07fdc561350ce" # Commit hash
+OUTPUT_BASE_FOLDER="./internal/capability/caldera/api"
+
+OUTPUT="$PROJ_PATH/caldera_api_raw.json"
+
+DOCKER_CMD="podman"
+# DOCKER_CMD="docker"
```
Comment (mpboom) [inline build_caldera_api.sh]: Either keep Docker the default or auto-discover it

File: build_caldera_api.sh [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+fi
+
+mkdir $TMP_WORKDIR
+cd $TMP_WORKDIR
+
+git clone $CALDERA_REPO caldera
+cd caldera/
+git checkout $CALDERA_VERSION
+git submodule init
+git submodule update --recursive
```
Comment (mpboom) [inline build_caldera_api.sh:33]: It feels a bit redundant to actually build the entire image... run it... then remove it again just to obtain the JSON. I get where this comes from, but would like to investigate if we can integrate the process a bit more - the user may want to use Caldera anyway! A good first step probably would be to rely (more) on Docker Compose for this process. I also think it's pretty in line with @MrDurion s comments, to see if we can rely less on scripts outside of Docker.

**Comment by mpboom** [comment]
> The modules imported in the generated code are not in the "required modules" yet (go.mod). Should this be added or is there a better convention to fix that?

I have no idea, but good and legitimate question

**Comment by mpboom** [comment]
> IMO this is going too far. I can't think of a use case where the **developer** wouldn't have access to bash + python3. The python dependency is also a temporary workaround until caldera fixes their API. The script already spawns an instance of CALDERA in Docker, doing it inside another container would just complicate things with no clear benefit. No special python libraries or packages are used either, just vanilla python and bash.

Agree with Stan we should not introduce dependencies. Also agree with Dawid that the Python dependency is pretty temporary. However, as stated in the code review, we should investigate if we can do more of this stuff _within_ the Caldera Docker container anyway.

Having said that, we should continue to use this branch as a method of testing/validating that this client works as expected. Can we make proper HTTP calls with it? Has that been tested?

File: build_caldera_api.sh (hunk 4 above)
Comment (daw1012345) [reply build_caldera_api.sh:33]: Important to note that this is meant to be able to be executed by a SOARCA *dev* (not a user) once in a blue moon if CALDERA changes their API.

File: build_caldera_api.sh (hunk 3 above)
Comment (daw1012345) [reply build_caldera_api.sh]: :(

File: build_caldera_api.sh (hunk 4 above)
Comment (mpboom) [reply build_caldera_api.sh:33]: Yeah, that's very true, it's mostly about the dev process... it makes this all "probably fine". I'm not too worried about the dependencies. But whereever we can stick it within the existing Docker process we should.

File: build_caldera_api.sh (hunk 3 above)
Comment (mpboom) [reply build_caldera_api.sh]: I mean Docker is the standard, just `alias docker podman` if you want to use podman 😂

Comment (mpboom) [reply build_caldera_api.sh]: (or auto-discover, is Podman is there just use it)

**Comment by daw1012345** [comment]
My opinion is as follows: this is a helper script, provided as a courtesy for future SOARCA devs. If they want, they can also manually provision CALDERA, fetch the swagger spec and run the Go command. I wrote this script to be able to fetch the swagger file easily with no previous CALDERA setup for any CALDERA version.

**Comment by mpboom** [comment]
> My opinion is as follows: this is a helper script, provided as a courtesy for future SOARCA devs. If they want, they can also manually provision CALDERA, fetch the swagger spec and run the Go command. I wrote this script to be able to fetch the swagger file easily with no previous CALDERA setup for any CALDERA version.

Fine by me, lets keep the scripts as is.

There are other comments open on this PR.

**Review by mpboom — APPROVED** [review]
